# 异步

[toc]

## Promise 都有哪些静态方法

- resolve() 返回给定值解析后的Promise对象
- reject() 返回带有拒绝原因的Promise对象
- all(iterable) 这个Promise的resolve回调执行是在所有输入的promise的resolve回调都结束，或者输入的iterable里没有promise了的时候。它的reject回调执行是，只要任何一个输入的promise的reject回调执行或者输入不合法的promise就会立即抛出错误，并且reject的是第一个抛出的错误信息。
- allSettled() 返回一个在所有给定的promise都已经fulfilled或rejected后的promise，并带有一个对象数组，每个对象表示对应的promise结果
- any(iterable) 只要一个执行成功就返回那个成功的promise对象，如果没有一个成功的promise对象，就返回一个失败的 promise 和AggregateError类型的实例，它是 Error 的一个子类，用于把单一的错误集合在一起。本质上，这个方法和Promise.all()是相反的。
- race(iterable) 返回一个 promise，一旦迭代器中的某个promise解决或拒绝，返回的 promise就会解决或拒绝。

## 如何终止Promise

Promise一旦创建不能取消，本质上Promise无法终止的

### 中断调用链

在then/catch最后返回一个永远pending状态的promise, 这样后面的then catch finally都不会执行

```js
return new Promise((resolve,reject)=>{})
```

### 中断Promise

合适的时候，把 pending 状态的 promise 给 reject 掉。例如一个常见的应用场景就是希望给网络请求设置超时时间，一旦超时就就中断

## async/await 的错误捕获[github](https://github.com/davidlin88/blogs/issues/2)

||try/catch|Golang风格|.catch()|
|:--:|:--:|:--:|:--:|
|优点|最保险的，可捕获同步、await的异步错误|流程控制准确;优雅|能捕获异步函数体内的同步和抛出的异步错误|
|缺点|无法捕获 try块中，被return 的异步错误（可用return await解决）;不能优雅地对错误分别处理;不适用于流程中不需要中断的错误;无法避免catch本身抛出异常且不好处理|大量重复 if(err !== null)，可能遗漏;无法捕获外部函数体内的同步错误|只适用于有 .catch() 方法的异步函数|
|场景|非常适合简单流程、不希望出错后影响后续执行流程的情景|适合链路长、较复杂的异步流程；适合于单个环节出错不影响流程|适合处理期望外的错误，及时处理；适合给.catch() 做兜底|

## 手写 Promise

## 实现异步请求的方式

```js
// 使用定时函数模拟异步请求：
// 方法一：使用回调函数
function doSomething(callback){
  setTimeout(function(){
    console.log('执行结束');
    let result = 4;
    callback(result);
  },100);
}

function callback(result){
  console.log('接收到结果为：'+result);
}

doSomething(callback);

// 方法二：promise对象
function doSomething(){
  return new Promise(function(resolve){
    setTimeout(function(){
      console.log('执行结束');
      let result = 6;
      resolve(result);
    },100);
  });

}

doSomething().then(result=>{
  console.log('接收到结果为：'+result);
});


// 方法三：generator函数
function doSomething(){
  setTimeout(function(){
    let result = 6;
    it.next(result);
  },100);
}

function* gener(){
  var result = yield doSomething();
  console.log(result);
}

let it = gener();
it.next();

// 方法四：async（ES7）
function doSomething(){
  return new Promise(resolve=>{
    setTimeout(function(){
      let result = 6;
      resolve(result);
    },100);
  });

}

async function action(){
  let result = await doSomething();
  console.log(result);
}
action();
```

## async / await 的原理是什么？

[掘金](https://juejin.cn/post/6844903988584775693)

## 如何顺序或并发执行 async / await

```js
/**
 * 间隔1s，依次打印10，20，30
 */
function Pro(val) {
  return new Promise((resolve, reject) => {
    setTimeout(function () {
      resolve(val);
    }, 1000);
  });
}

async function log(arrs) {
  for (const item of arrs) {
    var response = await Pro(item);
    console.log(response);
  }
}

log([10, 20, 30]);
```
### 给定一个url数组，如何实现接口的并发和继发

```js
// 继发实现
async function loadData(urls) {
  for (const url of urls) {
    const response = await fetch(url);
    console.log(await response.text());
  }
}

// 并发实现
async function loadData(urls) {
  // 并发读取 url
  const textPromises = urls.map(async url => {
    const response = await fetch(url);
    return response.text();
  });

  // 按次序输出
  for (const textPromise of textPromises) {
    console.log(await textPromise);
  }
}
```

## 如何实现大文件断点续传？

[掘金](https://juejin.cn/post/6844904046436843527)